feat:how-ai-sabotages-the-learning-curve-of-junior-developers (#93)
Ordina-Group/JWorks-blog: _posts/2024-11-08-how-ai-tools-sabotage-juniors-learning-curve.md

---
layout: post
authors: [ayoub_ait_cheikh_ahmed]
title: "How AI Sabotages the Learning Curve of Junior Developers"
image: /img/2024-11-08-how-ai-sabotages-juniors-learning-curve/banner.png
tags: [Junior Developer, AI, Learning Curve, Apprentice, Apprenticeship ]
category: Culture
comments: true
---

> > In coding, clean code is like the Krabby Patty secret formula—it’s all about the right ingredients for success!

Reviewer: Copy paste foutje 😉

> > In coding, clean code is like the Krabby Patty secret formula—it’s all about the right ingredients for success!

Reviewer: Copy paste foutje 😉

{:.no_toc}
- TOC
{:toc}

## Introduction

> This article stems from a moment of reflection during a drive home, where I pondered my apprenticeship and the learning process as a junior developer. 

Reviewer: Zou het niet beter zijn als je dat omgekeerd doet? Eerst kort introduceren en dan vertellen waarom je over dit onderwerp wil schrijven?

Using AI as a junior developer often provides a sense of hope and easy access to information, but this convenience comes with its own set of challenges when real (client) problems arise. In the spirit of preserving knowledge and fostering honest communication, I felt compelled to share my perspective on this journey.
This article stems from a moment of reflection during a drive home, where I pondered my apprenticeship and the learning process as a junior developer. This is aimed to comfort fellow juniors and newcomers and to offer relevant feedback to colleagues.

This is aimed to comfort fellow juniors and newcomers and to offer relevant feedback to colleagues.

Using AI as a junior developer often provides a sense of hope and easy access to information, but this convenience comes with its own set of challenges when real (client) problems arise.

> In the spirit of preserving knowledge and fostering honest communication, I felt compelled to share my perspective on this journey.

Reviewer: Zou het niet beter zijn als je dat omgekeerd doet? Eerst kort introduceren en dan vertellen waarom je over dit onderwerp wil schrijven 

Using AI as a junior developer often provides a sense of hope and easy access to information, but this convenience comes with its own set of challenges when real (client) problems arise. In the spirit of preserving knowledge and fostering honest communication, I felt compelled to share my perspective on this journey.

This article stems from a moment of reflection during a drive home, where I pondered my apprenticeship and the learning process as a junior developer. This is aimed to comfort fellow juniors and newcomers and to offer relevant feedback to colleagues.

The depth of these insights and experiences helps us retain information, carving it into stone.

In contrast, the quick, easy information provided by AI tools may feel useful at the moment but is often fleeting, like writing on the beach—easily erased by the next wave and quickly forgotten.
I've experienced the feeling of solving a problem with AI’s help only to realize later that I can't remember how or why it worked.
This led me to reflect on the types of information that stick and those that don’t.

### When AI is Incredibly Beneficial

AI can be highly advantageous when:

- **We need help with technologies outside our primary focus.**  
  For example, when I needed front-end code despite my primary focus being back-end development, AI tools provided fast and reliable assistance.
    This helps junior developers fill in knowledge gaps quickly, enabling them to move forward without getting stuck.

- **AI provides code snippets, templates, or solutions for unfamiliar areas.**  
  This can boost confidence and momentum, allowing us to progress through projects efficiently.

### The Drawbacks of Over-Reliance on AI

> 

Reviewer: Ik heb het gevoel dat stukje te kort is...

> 

Reviewer: Kan dit stukje niet wat uitgebreid?

2. **Ask Questions Without Fear**  
   Jake Scruggs, in *“My Apprenticeship at Object Mentor”*, highlighted: 

   > “Tomorrow I need to look stupider and feel better about it. This staying quiet and trying to guess what’s going on isn’t working so well.”

   Don’t hesitate to ask questions. Staying silent and guessing hinders your learning more than asking questions does.

3. **Identify and Fill Skill Gaps**  
   It’s normal to feel like others have skills you lack. Instead of being discouraged, pick one skill and commit to mastering it. By filling these gaps, you build a strong foundation.

4. **Embrace Challenges and Learn from Mistakes**  
   Mistakes are part of learning. Treat each one as a chance to improve your problem-solving skills and deepen your understanding.

5. **Code Reviews Are Compliments, Not Insults**  
   During my first code review, I felt disheartened by the feedback, especially since the codebase itself was flawed. However, I realized that code reviews are compliments—they signify someone investing time and expertise to help you grow.

## Balancing AI Assistance and Real Growth

> 

Reviewer: Zou dit deel niet bij de conclusie verwerkt kunnen worden? Het voelt als een goede samenvatting van de blogpost.

Misschien: 

AI holds immense potential to support junior developers by providing quick solutions and knowledge. However, it’s essential to use it wisely. Without balance, it can become a crutch that leaves you unprepared for challenges requiring deep, independent thinking. While AI can be a powerful tool, ensure it doesn’t overshadow the importance of developing strong analytical skills and comprehension. Quick fixes may solve immediate problems but won’t always contribute to lasting growth.

Apprenticeship is about facing knowledge gaps and nurturing problem-solving instincts. Embrace this journey with patience, and remember that every challenge adds to your expertise. By integrating AI responsibly and focusing on foundational skills, we can transform it from a tool that blindsides us to one that illuminates our path to development expertise. Reflect on your habits, adapt, and share these insights with peers to navigate the modern developer's journey together.

---
layout: post
authors: [ayoub_ait_cheikh_ahmed]
title: "How AI Sabotages the Learning Curve of Junior Developers"
image: /img/2024-11-08-how-ai-sabotages-juniors-learning-curve/banner.png

> tags: [Junior Developer, AI, Learning Curve, Apprentice, Apprenticeship ]

Reviewer 2: ```suggestion
tags: [AI, learning curve, apprentice, apprenticeship]
```

category: Culture
comments: true
---

> You could literally hammer a nail into a shoe, and it would get the job done. But does it really?

{:.no_toc}
- TOC
{:toc}

## Introduction

Using AI as a junior developer often provides a sense of hope and easy access to information, 
but this convenience comes with its own set of challenges when real (client) problems arise. 
In the spirit of preserving knowledge and fostering honest communication, 
I felt compelled to share my perspective on this journey.

This article stems from a moment of reflection during a drive home,
where I pondered my apprenticeship and the learning process as a junior developer. 
This is aimed to comfort fellow juniors and newcomers and to offer relevant feedback to colleagues.

<img alt="junior_reading" src="{{ '/img/2024-11-08-how-ai-sabotages-juniors-learning-curve/baby-reading.png' | prepend: site.baseurl }}" style="float: right; max-width: 180px;">

> ## Overview: The Dual Nature of AI in the Junior Developer Experience

Reviewer 2: Waarom de keuze van Overview?

Author: ik dacht mss een soort samenvatting die de lezer helpt om de section te begripen

> As junior developers, we are heavy consumers of information. 

Reviewer 2: Je spreekt vanuit jezelf, hoe kan je dan veralgemenen?

### When AI is Incredibly Beneficial

AI can be highly advantageous when:

- **We need help with technologies outside our primary focus.**  
  For example, when I needed front-end code despite my primary focus being back-end development, AI tools provided fast and reliable assistance.
  This helps junior developers fill in knowledge gaps quickly, enabling them to move forward without getting stuck.
   For instance, I was not familiar with Angular, but using AI to prompt me on how components work made it much easier to refactor and update my code.

- **AI provides code snippets, templates, or solutions for unfamiliar areas.**  
  AI can provide code snippets, templates, and solutions for areas that may be unfamiliar to developers. 
  This capability can significantly boost confidence and momentum, 
  enabling users to progress through projects more efficiently.
  For junior developers, this support can be particularly valuable when they encounter obstacles, 
  allowing them to overcome challenges and continue their work. 
 

### The Drawbacks of Over-Reliance on AI

Despite the numerous benefits that AI can offer, 

> over-dependence on these tools can lead to significant drawbacks that affect our long-term development and problem-solving capabilities.

Reviewer 2: een titel voor dit beetje. Dat is wel heel weinig. Misschien dan beter weglaten?

Secondly, making a habit of practicing without AI is key. 
Regularly writing and debugging code without AI assistance builds the foundational skills essential to any developer.
While AI can expedite tasks, the ability to manually construct, troubleshoot, and optimize code is irreplaceable. 
Practicing without AI fosters resilience and improves proficiency in critical areas like syntax fluency, 
logical structuring, and error handling. \
This hands-on experience is vital for developing a well-rounded skill set that can be relied upon in any situation.

Lastly, fostering an analytical spirit helps developers approach problems thoughtfully before turning to AI for solutions. 
Treat each coding challenge as an opportunity to independently think through potential solutions. 
By engaging deeply with the problem at hand, developers cultivate an analytical mindset that sharpens creativity and adaptability.
This approach is especially valuable in real-world development, where problems often demand innovative, 
context-sensitive solutions that AI alone may not adequately address.

## The Apprentice Mindset: A Foundation for Junior Developers

### What Does It Mean to Be an Apprentice?

> This is what I found defined in the book 
> *[Apprenticeship Patterns: Guidance for the Aspiring Software Craftsman](https://www.oreilly.com/library/view/apprenticeship-patterns/9780596806842/)* by Dave Hoover and Adewale Oshineye.
> 
> >Marten Gustafson, an expert we interviewed, summed up the apprentice mindset beautifully:
> >“I guess it basically means having the attitude that there’s always a better/smarter/faster way to do what you just did and what you’re currently doing. Apprenticeship is the state/process of evolving and looking for better ways and finding people, companies, and situations that force you to learn those better/smarter/faster ways.”

Reviewer 2: ```suggestion
>Marten Gustafson, an expert we interviewed, summed up the apprentice mindset beautifully:
>"I guess it basically means having the attitude that there’s always a better/smarter/faster way to do what you just did and what you’re currently doing. Apprenticeship is the state/process of evolving and looking for better ways and finding people, companies, and situations that force you to learn those better/smarter/faster ways."

This is what I found defined in the book 
*[Apprenticeship Patterns: Guidance for the Aspiring Software Craftsman](https://www.oreilly.com/library/view/apprenticeship-patterns/9780596806842/)* by Dave Hoover and Adewale Oshineye.
```

Using AI as a junior developer often provides a sense of hope and easy access to information, 
but this convenience comes with its own set of challenges when real (client) problems arise. 
In the spirit of preserving knowledge and fostering communication, 
I felt compelled to share my perspective on this journey.

This article comes from a moment of reflection during a drive home,
where I pondered my apprenticeship and the learning process as a junior developer. 
This is aimed to comfort fellow juniors and newcomers and to offer relevant feedback to colleagues.

<img alt="junior_reading" src="{{ '/img/2024-11-08-how-ai-sabotages-juniors-learning-curve/baby-reading.png' | prepend: site.baseurl }}" style="float: right; max-width: 180px;">

## Overview: The Dual Nature of AI in the Junior Developer Experience

> As junior developers, I am a heavy consumers of information. 

Reviewer 2: ```suggestion
As a junior developer, I consider myself a heavy consumer of information. 
```

### When AI is Incredibly Beneficial

AI proves highly advantageous in specific scenarios. 
For instance, when tackling technologies outside our primary focus, 
such as front-end development while specializing in back-end work, AI can offer fast, reliable guidance. 

This support helps junior developers bridge knowledge gaps quickly, enabling them to move forward without unnecessary delays.
 I found this particularly useful when working with Angular; 
 AI-assisted prompts helped me understand components and streamline my refactoring process.

Another benefit is AI’s ability to provide code snippets, templates, and solutions for unfamiliar areas. 

> These resources can boost confidence and momentum, allowing developers to progress more efficiently. For junior developers,
>  AI offers valuable support when they face obstacles, helping them overcome challenges and continue their work productively.

Reviewer 2: ```suggestion
These resources can boost confidence and momentum, allowing developers to progress more efficiently.
For junior developers, AI offers valuable support when they face obstacles, helping them overcome challenges and continue their work productively.
```

 
### The Drawbacks of Over-Reliance on AI

Despite the numerous benefits that AI can offer, 
over-dependence on these tools can lead to significant drawbacks that affect our long-term development and problem-solving capabilities.

AI-generated solutions provide immediate satisfaction and relief when faced with coding challenges.
However, the knowledge gained from these quick fixes is often fleeting.
When developers rely on AI without engaging in hands-on practice or active problem-solving,
 they miss the chance to solidify their understanding of the concepts. As a result, 
 information fades quickly, leaving only a vague recollection. 
This lack of retention can hinder the ability to tackle similar problems independently in the future.

> Additionally, heavy reliance on AI can create what i call **Dependency Addiction**.

Reviewer 2: ```suggestion
Additionally, heavy reliance on AI can create what I call **Dependency Addiction**.
```

 undercutting the advantages AI was intended to bring.

## Navigating AI as a Tool, Not a Crutch

Should we ride the AI wave? Yes, but with critical practices in place.

First, it's crucial to review and understand AI-generated code. 
Rather than accepting the code at face value, take the time to dissect it and grasp how it functions and why certain choices were made. 
This examination transforms AI outputs into valuable learning opportunities, 
allowing developers to deepen their understanding of specific patterns and practices. 
By questioning and analyzing AI’s suggestions, we can integrate these insights into our own problem-solving toolkit, 
enhancing our knowledge and making us more effective in coding independently.

Secondly, making a habit of practicing without AI is key. 
Regularly writing and debugging code without AI assistance builds the foundational skills essential to any developer.
While AI can expedite tasks, the ability to manually construct, troubleshoot, and optimize code is irreplaceable. 
Practicing without AI fosters resilience and improves proficiency in critical areas like syntax fluency, 

> logical structuring, and error handling. \

Reviewer 2: ```suggestion
logical structuring, and error handling.
```

This hands-on experience is vital for developing a well-rounded skill set that can be relied upon in any situation.

Lastly, fostering an analytical spirit helps developers approach problems thoughtfully before turning to AI for solutions. 
Treat each coding challenge as an opportunity to independently think through potential solutions. 
By engaging deeply with the problem at hand, developers cultivate an analytical mindset that sharpens creativity and adaptability.
This approach is especially valuable in real-world development, where problems often demand innovative, 
context-sensitive solutions that AI alone may not adequately address.

## The Apprentice Mindset: A Foundation for Junior Developers

### What Does It Mean to Be an Apprentice?

This is what I found defined in the book 
*[Apprenticeship Patterns: Guidance for the Aspiring Software Craftsman](https://www.oreilly.com/library/view/apprenticeship-patterns/9780596806842/)* by Dave Hoover and Adewale Oshineye.

>Marten Gustafson, an expert we interviewed, summed up the apprentice mindset beautifully:

> >“I guess it basically means having the attitude that there’s always a better/smarter/faster way to do what you just did and what you’re currently doing. Apprenticeship is the state/process of evolving and looking for better ways and finding people, companies, and situations that force you to learn those better/smarter/faster ways.”

Reviewer 2: ```suggestion
>Marten Gustafson, an expert we interviewed, summed up the apprentice mindset beautifully:
>"I guess it basically means having the attitude that there’s always a better/smarter/faster way to do what you just did and what you’re currently doing. Apprenticeship is the state/process of evolving and looking for better ways and finding people, companies, and situations that force you to learn those better/smarter/faster ways."

This is what I found defined in the book 
*[Apprenticeship Patterns: Guidance for the Aspiring Software Craftsman](https://www.oreilly.com/library/view/apprenticeship-patterns/9780596806842/)* by Dave Hoover and Adewale Oshineye.
```
